kernel/aster-nix/src/sched/priority_scheduler.rs

```diff
@@ -17,40 +22,226 @@ pub fn init() {
 
-impl PreemptScheduler {
+impl<T: PreemptSchedInfo> PreemptRunQueue<T> {
     pub fn new() -> Self {
         Self {
-            real_time_tasks: SpinLock::new(LinkedList::new(TaskAdapter::new())),
-            normal_tasks: SpinLock::new(LinkedList::new(TaskAdapter::new())),
+            current: None,
+            real_time_tasks: VecDeque::new(),
+            normal_tasks: VecDeque::new(),
         }
     }
 }
 
-impl Scheduler for PreemptScheduler {
-    fn enqueue(&self, task: Arc<Task>) {
-        if task.is_real_time() {
-            self.real_time_tasks.lock_irq_disabled().push_back(task);
-        } else {
-            self.normal_tasks.lock_irq_disabled().push_back(task);
+impl<T: Sync + Send + PreemptSchedInfo> LocalRunQueue<T> for PreemptRunQueue<T> {
+    fn current(&self) -> Option<&Arc<T>> {
+        self.current.as_ref().map(|entity| &entity.runnable)
+    }
+
+    fn update_current(&mut self, flags: UpdateFlags) -> bool {
+        match flags {
+            UpdateFlags::Tick => {
+                if self.current.is_none() {
+                    return false;
+                }
+                self.current.as_mut().unwrap().ran_out_of_time()
+                    || (!self.current.as_ref().unwrap().is_real_time()
+                        && !self.real_time_tasks.is_empty())
+            }
+            _ => true,
         }
     }
 
-    fn dequeue(&self) -> Option<Arc<Task>> {
-        if !self.real_time_tasks.lock_irq_disabled().is_empty() {
-            self.real_time_tasks.lock_irq_disabled().pop_front()
+    fn pick_next_current(&mut self) -> Option<Arc<T>> {
+        let prev = self.current.take();
+        let next = if !self.real_time_tasks.is_empty() {
+            self.real_time_tasks.pop_front()
         } else {
-            self.normal_tasks.lock_irq_disabled().pop_front()
+            self.normal_tasks.pop_front()
+        };
+        self.current = next.clone();
+        if self.current.is_none() {
+            self.current = prev;
+        } else if let Some(prev_entity) = prev {
+            if prev_entity.is_real_time() {
+                self.real_time_tasks.push_back(prev_entity);
+            } else {
+                self.normal_tasks.push_back(prev_entity);
+            }
+        }
+
+        next.map(|entity| entity.runnable)
+    }
+
+    fn dequeue_current(&mut self) -> Option<Arc<T>> {
+        let current = self.current.take().map(|entity| entity.runnable);
+        current.as_ref().unwrap().remove_from_cpu();
+
+        current
+    }
+}
+
+struct PreemptSchedEntity<T: PreemptSchedInfo> {
+    runnable: Arc<T>,
+    time_slice: TimeSlice,
+}
+
+impl<T: PreemptSchedInfo> PreemptSchedEntity<T> {
+    fn new(runnable: Arc<T>) -> Self {
+        Self {
+            runnable,
+            time_slice: TimeSlice::default(),
+        }
+    }
+
+    fn is_real_time(&self) -> bool {
+        self.runnable.is_real_time()
+    }
+
+    fn ran_out_of_time(&mut self) -> bool {
+        self.time_slice.ran_out()
```

Comment:
The name `ran_out` or `ran_out_of_time` looks to be an accessor method returning whether the test has run out. It turns out that these methods also increase the elapsed time. 

I suggest calling this function `tick(&self)` or `elapse(&self, ticks: u32)` and rename the `count` field of `TimeSlice` to `elapsed_ticks`.